## 🔍 Build Failure Analysis

> **Workflow run**: https://github.com/microsoft/testfx/actions/runs/26975085514

### ❌ Root Cause: `IDE0028` — Collection initialization can be simplified

The build fails with **3 errors** (one per target framework: `netstandard2.0`, `net8.0`, `net9.0`), all pointing to the **same single location**:

```
src/Platform/Microsoft.Testing.Extensions.VSTestBridge/ObjectModel/ObjectModelConverters.cs(145,56): error IDE0028
```

The newly introduced line:
```csharp
List<TrxMessage>? trxMessages = isTrxEnabled ? new List<TrxMessage>(testResult.Messages.Count) : null;
```

triggers IDE0028 ("Collection initialization can be simplified") because `.editorconfig` sets `dotnet_style_prefer_collection_expression = true` and the type `List<TrxMessage>` is redundant — it can be inferred from the left-hand side. In CI the project builds with `/p:TreatWarningsAsErrors=true`, promoting this warning to an error.

### ✅ Fix

Use a **target-typed `new` expression** to drop the redundant type annotation while preserving the capacity hint:

```csharp
// Before (triggers IDE0028)
List<TrxMessage>? trxMessages = isTrxEnabled ? new List<TrxMessage>(testResult.Messages.Count) : null;

// After
List<TrxMessage>? trxMessages = isTrxEnabled ? new(testResult.Messages.Count) : null;
```

This is the minimal one-character change that satisfies the analyzer, keeps the pre-sizing optimisation intact (important on `netstandard2.0` where `EnsureCapacity` is unavailable), and is idiomatic C#.

An inline suggestion has been posted directly on the offending line below.



> Generated by [Build Failure Analysis](https://github.com/microsoft/testfx/actions/runs/26975085514) for issue #8823 · sonnet46 2.6M · [◷](https://github.com/search?q=repo%3Amicrosoft%2Ftestfx+%22gh-aw-workflow-call-id%3A+microsoft%2Ftestfx%2Fbuild-failure-analysis%22&type=issues)

<!-- gh-aw-agentic-workflow: Build Failure Analysis, engine: copilot, version: 1.0.52, model: claude-sonnet-4.6, id: 26975085514, workflow_id: build-failure-analysis, run: https://github.com/microsoft/testfx/actions/runs/26975085514 -->
<!-- gh-aw-workflow-call-id: microsoft/testfx/build-failure-analysis -->

--- a/src/Platform/Microsoft.Testing.Extensions.VSTestBridge/ObjectModel/ObjectModelConverters.cs
+++ b/src/Platform/Microsoft.Testing.Extensions.VSTestBridge/ObjectModel/ObjectModelConverters.cs
@@ -138,6 +138,44 @@ internal static class ObjectModelConverters
 
         testNode.AddOutcome(testResult);
 
+        // Validate TRX prerequisites (and add TRX-only properties that don't depend on the message
+        // loop) BEFORE iterating testResult.Messages, so that an InvalidOperationException from a
+        // malformed FullyQualifiedName still wins over the UnreachableException that the message
+        // loop can throw on an unexpected msg.Category. This preserves the pre-PR exception order.
+        if (isTrxEnabled)
+        {
+            if (!RoslynString.IsNullOrEmpty(testResult.ErrorMessage) || !RoslynString.IsNullOrEmpty(testResult.ErrorStackTrace))
+            {
+                testNode.Properties.Add(new TrxExceptionProperty(testResult.ErrorMessage, testResult.ErrorStackTrace));
+            }
+
+            testNode.Properties.Add(new TrxTestDefinitionName(testResult.TestCase.DisplayName ?? testResult.TestCase.FullyQualifiedName));
+
+            TestMethodIdentifierProperty? testMethodIdentifierProperty = testNode.Properties.SingleOrDefault<TestMethodIdentifierProperty>();
+            if (testMethodIdentifierProperty is not null)
+            {
+                // TODO: Should TRX className have arity for generic classes?
+                if (RoslynString.IsNullOrEmpty(testMethodIdentifierProperty.Namespace))
+                {
+                    testNode.Properties.Add(new TrxFullyQualifiedTypeNameProperty(testMethodIdentifierProperty.TypeName));
+                }
+                else
+                {
+                    testNode.Properties.Add(new TrxFullyQualifiedTypeNameProperty($"{testMethodIdentifierProperty.Namespace}.{testMethodIdentifierProperty.TypeName}"));
+                }
+            }
+            else if (TryParseFullyQualifiedType(testResult.TestCase.FullyQualifiedName, out string? fullyQualifiedType))
+            {
+                // VSTest's TestCase.FQN is very non-standard.
+                // We should avoid using it if we can, and so we prefer TestMethodIdentifierProperty first.
+                testNode.Properties.Add(new TrxFullyQualifiedTypeNameProperty(fullyQualifiedType));
+            }
+            else
+            {
+                throw new InvalidOperationException("Unable to parse fully qualified type name from test case: " + testResult.TestCase.FullyQualifiedName);
+            }
+        }
+
         // Single pass over testResult.Messages: collect TRX messages, standard-error, and
         // standard-output in one loop, avoiding a separate LINQ Select + spread + foreach.
         // Pre-size trxMessages to the exact capacity to avoid List<T> growth reallocations
@@ -173,37 +211,6 @@ internal static class ObjectModelConverters
 
         if (isTrxEnabled)
         {
-            if (!RoslynString.IsNullOrEmpty(testResult.ErrorMessage) || !RoslynString.IsNullOrEmpty(testResult.ErrorStackTrace))
-            {
-                testNode.Properties.Add(new TrxExceptionProperty(testResult.ErrorMessage, testResult.ErrorStackTrace));
-            }
-
-            testNode.Properties.Add(new TrxTestDefinitionName(testResult.TestCase.DisplayName ?? testResult.TestCase.FullyQualifiedName));
-
-            TestMethodIdentifierProperty? testMethodIdentifierProperty = testNode.Properties.SingleOrDefault<TestMethodIdentifierProperty>();
-            if (testMethodIdentifierProperty is not null)
-            {
-                // TODO: Should TRX className have arity for generic classes?
-                if (RoslynString.IsNullOrEmpty(testMethodIdentifierProperty.Namespace))
-                {
-                    testNode.Properties.Add(new TrxFullyQualifiedTypeNameProperty(testMethodIdentifierProperty.TypeName));
-                }
-                else
-                {
-                    testNode.Properties.Add(new TrxFullyQualifiedTypeNameProperty($"{testMethodIdentifierProperty.Namespace}.{testMethodIdentifierProperty.TypeName}"));
-                }
-            }
-            else if (TryParseFullyQualifiedType(testResult.TestCase.FullyQualifiedName, out string? fullyQualifiedType))
-            {
-                // VSTest's TestCase.FQN is very non-standard.
-                // We should avoid using it if we can, and so we prefer TestMethodIdentifierProperty first.
-                testNode.Properties.Add(new TrxFullyQualifiedTypeNameProperty(fullyQualifiedType));
-            }
-            else
-            {
-                throw new InvalidOperationException("Unable to parse fully qualified type name from test case: " + testResult.TestCase.FullyQualifiedName);
-            }
-
             testNode.Properties.Add(new TrxMessagesProperty(trxMessages is { Count: > 0 } ? [.. trxMessages] : []));
         }
 

--- a/src/Platform/Microsoft.Testing.Extensions.VSTestBridge/ObjectModel/ObjectModelConverters.cs
+++ b/src/Platform/Microsoft.Testing.Extensions.VSTestBridge/ObjectModel/ObjectModelConverters.cs
@@ -142,7 +142,9 @@ internal static class ObjectModelConverters
         // standard-output in one loop, avoiding a separate LINQ Select + spread + foreach.
         // Pre-size trxMessages to the exact capacity to avoid List<T> growth reallocations
         // (every entry in testResult.Messages produces one TRX message when TRX is enabled).
+#pragma warning disable IDE0028 // Collection initialization can be simplified - capacity hint is intentional.
         List<TrxMessage>? trxMessages = isTrxEnabled ? new List<TrxMessage>(testResult.Messages.Count) : null;
+#pragma warning restore IDE0028
         List<string>? standardErrorMessages = null;
         List<string>? standardOutputMessages = null;
         foreach (TestResultMessage msg in testResult.Messages)

